text/0088-metric-instrument-optional-refinements.md

```diff
@@ -0,0 +1,205 @@
+measurements of changes in a sum, therefore it uses `Add()` instead of
+`Record()`, and it specifies `Sum` as the standard aggregation.
+
+What this illustrates is that we have modeled this space poorly.  This
+does not propose to change any existing metrics APIs, only our
+understanding of the three instruments currently included in the
+specification: Measure, Observer, and Counter.
+
+## Explanation
+
+The Measure and Observer instrument are defined as _foundational_
+here, in the sense that any kind of metric instrument must reduce to
+one of these.  The foundational instruments are unrestricted, in the
+sense that metric events support any numerical value, positive or
+negative, zero or infinity.
+
+The distinction between the two foundational instruments is whether
+they are synchronous.  Measure instruments are called synchronously by
+the user, while Observer instruments are called asynchronously by the
+implementation.  Synchronous instruments (Measure and its refinements)
+have three calling patterns (_Bound_, _Unbound_, and _Batch_) to
+capture measurements.  Asynchronous instruments (Observer and its
+refinements) use callbacks to capture measurements.
+
+All measurement APIs produce metric events consisting of [timestamp,
+instrument descriptor, label set, and numerical
+value](api-metrics.md#metric-event-format).  Synchronous instrument
+events additionally have [Context](api-context.md), describing
+properties of the associated trace and distributed correlation values.
+
+Observer instruments have a well-defined _last value_ measured by the
+instrument that can be useful in defining aggregations.  To maintain
+this property, we impose this requirement: two or more calls to
+`Observe()` in a single Observer callback invocation are treated as
+duplicates of each other, and the last call to `Observe()` wins.
+
+Measure instruments do not define a _last value_ relationship.
+
+### Standard implementation of Measure and Observer
+
+OpenTelemetry specifies how the default SDK should treat metric
+events, by default, when asked to export data from an instrument.
+Measure and Observer instruments compute `Sum` and `Count`
+aggregations, by default, in the standard implementation.  This pair
+of measurements, of course, defines an average value.  There are no
+restrictions placed on the numerical value in an event for the two
+foundational instruments.
+
+### Refinements to Measure and Observer
+
+The `Monotonic` and `Absolute` options were removed in the 0.3
+specification.  Here, we propose to regain the equivalent effects
+through instrument refinements.  Instrument refinements have the same
+calling patterns as the foundational instrument they refine, adding
+either a different standard implementation or a restriction of the
+input domain.
+
+We have done away with instrument options, in other words, in favor of
+optional metric instruments.  Here we discuss three important
+capabilities achievable using instrument refinements.
```

Comment:
One open question: should we allow users to create these base instruments or do they serve as "abstract" classes for all the other instruments?

For Measure instruments we can have (names not well thought):
* Counter - monotonic Sum.
* UpDownCounter - non-monotonic Sum.
* Distribution - min/max/sum/count (or just sum/count) - only positive values.
* DistributionWithNegative - min/max/sum/count (or just sum/count) - can have negative values.

It is important to differentiate between Distribution with negative and non-negative to be compatible with OpenMetrics (this is unfortunately).

For Observer (names match a bit the Prometheus/Stackdriver terminology and may not be the best names):
* GaugeObserver - instantaneous measurement of a value (values for the same labels can go up down)
* CumulativeObserver - measurements for the same labels are going up.
* DeltaObserver - measurements for the same labels represent the change since previous observer call, and can only be positive).

The main idea for my comment is the first sentence.

Reply:
I was leaning toward allowing the base instruments to be non-abstract, i.e., `DistributionWithNegative` is the same as `Measure`.

Reply:
The only downside is that these generic names do not express that way the semantic as the other names I was proposing. Not saying I cannot live with that but just an idea.

Reply:
I was leaning the other way, but I re-read your proposal above and am starting to like it, except...

For synchronous instruments, I'm not convinced that it matters if a `Distribution` is non-negative. If we allow that `Distribution` is unrefined (i.e., positive or negative), there's no need for a `DistributionWithNegative`. Do you have any particular reasons why we care that Distribution values are non-negative? I have some in mind, but I don't think they matter. If you want to sum the values and compute a rate, a Counter configured with an alternate aggregation would be an alternative (i.e., using a View API).

For the asynchronous instruments you listed, I might replace `GaugeObserver` with `LastValueObserve`.

Then the proposal at the end of the latest draft of this PR would propose these non-abstract instruments:

Synchronous:
```
Counter
UpDownCounter
Timing
Distribution
```

Asynchronous:
```
DeltaObserver
CumulativeObserver
LastValueObserver
```

Reply:
1. I would keep `Timing` separately because that can be implemented as just a measurement 
 helper class or can be an instrument, here is an example for measurement helper class:
    ```
    t := Timer(distribution, labels);
    ...  // your code
    t.Record();
    ```
    OR as a standalone instrument.

1. For the negative distribution as I mentioned it is unfortunate that Prometheus (when exported as a histogram) uses different ways to export the count/sum uses `_count/_sum` or `_gcount/_gsum` for histogram with negative values https://github.com/prometheus/client_python/blob/master/prometheus_client/openmetrics/parser.py#L392 so we kind of have to be able to support that. Maybe we can start initially with `Distribution` only support positive values and later after we understand better relax that condition or add a `NonNegativeDistribution`.

1. Does DeltaObserver and CumulativeObserver imply that the raw values used when do the pre-aggregation are positive only (that would be great because will match perfectly with a Counter Sum aggregation that we can produce and LastValueObserver will match perfectly with a Sum aggregation for a UpDownCounter)?

Reply:
(1) Yes, but are we willing to specify this helper class?  It feels like an inconsistency to say that timing measurements should be made using a different class of thing, compared with non-timing measurements. One of the key ideas is that the units are automatically specified for Timing measurements, and in the snippet of code above this could break. How can we avoid using a Timing wrapper with a Distribution instrument that has the wrong units?

(2) Ack. I'm skeptical that `NonNegativeDistribution` is an important class of instrument. Users could configure a `Counter` instrument to be aggregated as a distribution and get the same effect (a distribution of strictly non-negative values), but that would be a non-standard configuration. I have an internal debate about this: when the former Prometheus Histogram was used **correctly**, a Distribution will **correctly** replace it.  If the former Prometheus Histogram's error behavior is taken into account, a Distribution is not a correct replacement. I would argue we have a compatible instrument, in this case. I do not believe error-compatibility should be a requirement for OTel. 

The problem here is that for new users of a Distribution instrument, we don't know whether to export as a Prometheus counter-histogram or gauge-histogram. A Prometheus exporter could be configured to export one distribution for non-negative values and one distribution for negative values. Users _ought_ to be able to combine the positive and negative distributions correctly downstream--in any case IMO this is not an error at the API level, this is an incompatibility in a legacy export path. I do not believe we should let Prometheus shape the OTel API to this extent.

(3) Yes, my intent was that `DeltaObserver` and `CumulativeObserver` accept non-negative values. The `UpDown-` variation on these would support negative values. I see there's a minor complication here, but I've waved this away already in the document: I don't think it matters whether a Cumulative instrument accepts non-negative values, what we care about is that it has a non-negative rate. I'm willing to say that `CumulativeObserver` has both non-negative and non-negative-rate refinements. I wrote:

> For the precomputed-sum instruments, we technically do not care whether the inputs are non-negative, because rate aggregation computes differences.  However, it is useful for other aggregations to assume that precomputed sums start at zero, and we will ignore the case where a precomputed sum has an initial value other than zero.